Extract mutable builder from `ModuleScope`

hubertp opened the pull request:
### Pull Request Description

Refactored mutable parts of `ModuleScope` into builder to make it easier to reduce unnecessary locks.

### Important Notes

Elements of ModuleScope (types, imports etc) are used while _building_ of it may still be in progress. In order to make static typing happy, every `ModuleScope.Builder` can be exposed as (unmodifiable) `ModuleScope`.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

hubertp committed:
Minor refactorings

hubertp committed:
wip

hubertp committed:
don't reset types in scopes

hubertp committed:
Make builders usage explicit

hubertp committed:
minor allocation optimizations

hubertp committed:
reduce lock scope

hubertp committed:
remove redundant DelayedModuleScope

hubertp committed:
nits

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java:
```diff
@@ -28,44 +28,28 @@
-  private Map<Type, Map<Type, Function>> conversions;
-  private Set<ModuleScope> imports;
-  private Set<ModuleScope> exports;
+  private final Map<String, Object> polyglotSymbols;
```
Yup, this is great. Having all fields in `ModuleScope` final, is the goal.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java:
```diff
@@ -276,30 +155,12 @@ private Function getExportedConversion(Type type, Type target) {
-  }
-
   public Map<String, Type> getTypes() {
     return types;
```
We should make sure the returned `Map` is _immutable_. E.g. wrap it by `Collections.unmodifiableMap` either here or in the constructor.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java:
```diff
@@ -461,4 +262,270 @@ Type getType() {
+
+    public ModuleScope build() {
+      if (moduleScope == null) {
+        synchronized (this) {
```
`synchronized`? Only useful if we encourage the `Builder` to be used from multiple threads. But it shouldn't be used that way. A `Builder` should only be used from a single thread, am I right?

If so, we should `assert Thread.currentThread() == builderCreatorThread` at few places to ensure `Builder` is always used from a single thread.

Akirathan replied:
AFAIK, `ModuleScope` is constructed only from `IrToTruffle` and that is single-threaded. Moreover, only one module is processed at single time. So I believe that usage of `synchronized` does not make sense as well. Moreover, I believe that it does not make sense to use `ConcurrentHashMap` as fields.

Akirathan replied:
Note that I have removed all the `ConcurrentHashMap` stuff in my recently [merged PR](https://github.com/enso-org/enso/pull/9844#discussion_r1609807230).

hubertp replied:
If we enable multi-threading for visualizations, which can as a side-effect trigger compilation, then we can have multiple threads accessing/adding to the builder. That was also the motivation for this PR, to make it more explicit.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java:
```diff
@@ -461,4 +262,270 @@ Type getType() {
+      conversions = new ConcurrentHashMap<>();
+      imports = new HashSet<>();
+      exports = new HashSet<>();
+      moduleScope = null;
```
This is inconsistent with the synchronized block in `build` - make whole `reset` synchronized? Remove synchronization and _assert only single threaded usage_ of the builder?

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/interpreter/runtime/IrToTruffle.scala:
```diff
@@ -131,7 +131,7 @@ import scala.jdk.OptionConverters._
   val context: EnsoContext,
   val source: Source,
-  val moduleScope: ModuleScope,
+  val moduleScope: ModuleScope.Builder,
```
That looks promising!

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/EnsoRootNode.java:
```diff
@@ -145,10 +143,18 @@ public LocalScope getLocalScope() {
    * @return the module scope for this node
    */
-  public ModuleScope getModuleScope() {
+  public ModuleScope.Builder getModuleScope() {
```
Ideally such a `public` getter of the builder should be avoided. The `Builder` should be `private` to the creator of it, not floating around being accessed by random parties.

hubertp replied:
Yes, I was hoping that would be the case as well, except that we are currently dealing with a situation when the builder's state is also inspected while stuff is added to it.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/CaptureCallerInfoNode.java:
```diff
@@ -55,7 +55,7 @@ public LocalScope getLocalScope() {
     }
 
-    public ModuleScope getModuleScope() {
+    public ModuleScope.Builder getModuleScope() {
```
Why does `ScopeInfo` needs a `Builder`? Shouldn't all the scope already be built before execution even gets here?

hubertp replied:
Yes, I know this one is not obvious but this one is needed because of `EvalNode`'s `parseExpression` which builds `ClosureRootNode`, which in turn requires the builder because that's the only thing that `IrToTruffle` has when it creates those as well. catch-22.

JaroslavTulach reviewed (approved):
Ideally, I'd like the `Builder` to be more hidden, not flying around being accessed by everybody. But even the current state that makes `ModuleScope` immutable is a step forward and deserves to be integrated. Thank you!

Akirathan commented on engine/runtime-integration-tests/src/test/java/org/enso/interpreter/node/expression/builtin/meta/TypeOfNodeTest.java:
```diff
@@ -56,7 +57,7 @@ public static Object[][] allPossibleEnsoInterpreterValues() throws Exception {
       }
     }
-    data.add(new Object[] {UnresolvedSymbol.build("unknown_name", null), "Function"});
+    data.add(new Object[] {UnresolvedSymbol.build("unknown_name", (ModuleScope) null), "Function"});
```
Why is this change necessary? Casting `null` to `ModuleScope`?

hubertp replied:
`build` is currently overloaded.

Akirathan commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java:
```diff
@@ -461,4 +262,270 @@ Type getType() {
+      Set<ModuleScope.Builder> imports = new HashSet<>(this.imports);
+      Set<ModuleScope.Builder> exports = new HashSet<>(this.exports);
+      this.types
+          .entrySet()
```
```suggestion
          .forEach((key, value) -> {
            if (typeNames.contains(key)) {
              requestedTypes.put(key, value);
            }
          });
```

JaroslavTulach replied:
`for (var en : types.entrySet()` is easier to step thru in debugger.

Akirathan replied:
When you apply my suggestion, you don't have to call `entry.getKey()` and `entry.getValue()` in the subsequent code. That was the main reason for the suggestion.

Akirathan commented on engine/runtime/src/main/java/org/enso/interpreter/node/ClosureRootNode.java:
```diff
@@ -54,7 +54,7 @@ public class ClosureRootNode extends EnsoRootNode {
       EnsoLanguage language,
       LocalScope localScope,
-      ModuleScope moduleScope,
+      ModuleScope.Builder moduleScope,
```
I believe only `IrToTruffle` should use `ModuleScope.Builder` for its purposes. All the runtime nodes should have only immutable `ModuleScope`. Is that possible? Having `ModuleScope.Builder` as fields in some truffle nodes diminishes the whole idea of this PR, since we will still be able to mutate the underlying builder at runtime, which should be restricted. No?

Akirathan commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Type.java:
```diff
@@ -28,8 +28,9 @@
+
   private final String name;
-  private @CompilerDirectives.CompilationFinal ModuleScope definitionScope;
+  private @CompilerDirectives.CompilationFinal ModuleScope.Builder definitionScope;
```
DOes it make sense to declare fields of data as `@CompilationFinal`?

JaroslavTulach replied:
It might make sense. Instances of `Type` are directly referenced from various Enso `Node` - e.g. they are part of the AST. Remember we have the [EXCLUSIVE](https://github.com/enso-org/enso/blob/202f7e1a4950fe77b4c7bb746e83c98e7d256750/engine/runtime/src/main/java/org/enso/interpreter/EnsoLanguage.java#L70) sharing policy - there is no sharing of ASTs.

JaroslavTulach replied:
However, please don't reference _builder_, reference `Module`.

Akirathan reviewed (changes requested):
Altough draft PR, I am still submitting "Request changes" here. In the current form, I don't see a lot of improvements over the previous state. I would like to see most, if not all, the runtime nodes and classes to receive immutable `ModuleScope` as parameters and fields, not `ModuleScope.Builder`.

hubertp committed:
Merge branch 'develop' into wip/hubert/9736-visualizations

hubertp commented:
@Akirathan @JaroslavTulach 
To cut the long story short, your requirement of having only `ModuleScope` is satisfied in runtime because you won't be able to mutate builder once something calls `build()`. Statically, yes, you still see `ModuleScope.Builder` in various places.

I understand the sentiment that builder shouldn't be _visible_ outside of `IrToTruffle` but take into account that since Truffle nodes are built during `IrToTruffle` the only thing that I can provide reference to is `ModuleScope.Builder` not `ModuleScope`. 
In fact that is pretty much the only reason why you still see `ModuleScope.Builder` in various nodes.
I wanted to have only `ModuleScope` in Truffle nodes but I have yet to figure out, if possible at all, how to make typechecker happy with that requirement.

hubertp commented on engine/runtime/src/main/scala/org/enso/interpreter/runtime/IrToTruffle.scala:
```diff
@@ -740,6 +724,7 @@ class IrToTruffle(
         moduleScope.registerConversionMethod(toType, fromType, function)
       }
     })
+    moduleScope.build()
```
Maybe this line got lost in the amount of changes but it is probably worth putting it in **bold**. This ensures that `ModuleScope.Builder` can't be modified.

hubertp committed:
Illustrate when immutable ModulesScope is created

Added `built` method, apart from the existing `build` method, to return
ModuleScope associated with the builder.
This way it becomes clear when builder transforms into `ModuleScope`.

JaroslavTulach commented:
> builder shouldn't be _visible_ outside of `IrToTruffle` but take into account that since Truffle nodes are built during `IrToTruffle` the only thing that I can provide reference to is `ModuleScope.Builder` not `ModuleScope`. 

Don't we have to support _some form of mutability_ to allow recompilation of `Module`? Where such _mutability_ happens? A `Module` may get assigned different `ModuleScope`? If so...

> I can provide reference to is `ModuleScope.Builder` not `ModuleScope`.

... provide reference to uninitialized `Module` and fill it in later when `.build()` is called.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java:
```diff
@@ -53,18 +53,18 @@
-  private ModuleScope scope;
   private ModuleSources sources;
+  private QualifiedName name;
+  private final ModuleScope.Builder scopeBuilder;
```
Reference to builder - eh, I really cannot make any (static) judgements about such a design. Having a builder here doesn't make any logical sense to me.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java:
```diff
@@ -385,10 +384,9 @@ public final boolean isModuleSource(Source s) {
     Source source = getSource();
     if (source == null) return;
-    scope.reset();
+    scopeBuilder.reset();
```
Is the builder really immutable once `.build()`? What does `.reset()` then do?

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java:
```diff
@@ -460,4 +265,282 @@ Type getType() {
+
+    public void reset() {
+      polyglotSymbols = new LinkedHashMap<>();
+      // can't clear types because on recompilation methods etc will be assigned to the new one
```
As suggested elsewhere, I would use `Module` as _eventually consistent_ view of `ModuleScope`. I'd add:
```java
class Module {
  /** may return different instance of scope (when reparsing happened), but always consistenly filled from the latest successfull (re)parse.
  ModuleScope getModuleScope();

  /** Start building new {@link #getModuleScope()} for this Module. Records current module state for later
  * {@code build()} call on the builder.
  * 
 * @return new builder to create and set new ModuleScope into this Module. Reuses references to existing {@code types}.
 */
  ModuleScope.Builder newBuilder();
}

/** Creates new {@code ModuleScope} and atomically sets it to associated {@code Module}. Builder is made
* unusable after calling its {@code build()} method.
* @throws ConcurrentModificationException if the {@code Module} was modified since the creation of the builder by some other builder
*/
void ModuleScope.Builder.build();

hubertp replied:
I recall why I couldn't do it this way - when constructing truffle nodes (`IrToTruffle`) we occasionally reference modules. Those modules are still wip, meaning that their `ModuleScope` is not available, yet, but we need it to resolve methods. I will still try to figure out if we can get something vaguely similar to the above requirement.

JaroslavTulach replied:
> Those modules are still wip, meaning that their ModuleScope is not available

What do you mean by wip? 
- is `Module` instance available? Then keep pointer to it, it will get `ModuleScope` eventually...
- ok, `ModuleScope` isn't yet ready, but it will be eventually
- _"we need it to resolve methods"_ - how can you resolve methods without finishing resolution of `ModuleScope`?

Is the code resolving methods based on partially resolved `ModuleScope`? That seems buggy to even consider...

hubertp replied:
> Is the code resolving methods based on partially resolved ModuleScope? That seems buggy to even consider...

I don't see any gatekeepers preventing it, so yeah.

hubertp committed:
Introduce ImportExportScope proxy

`ImportExportScope` serves as a proxy to the underlying module's scope.
Can't reference the scope directly at the point when it is built because
building of some modules might still be in progress.

This also allowed to get rid off the horrible `withTypes` method that
would create a temporary module scope.

hubertp committed:
Ditch `get_unresolved_symbol_scope`

It's unused and untested anyway.

hubertp committed:
Drop built/build check for ModuleScope.Builder

Simplified the whole approach by avoiding calls to `built` and `build`
to determine the state of `ModuleScope.Builder`.
Instead, `ModuleScope.Builder` is only passed whenever a registration is
necessary. Since builder is accessed in various places, while it may
still, in theory, be constructor, this change introduces a simple,
short-lived, proxy that delegates to accessors and satisfies the
`ModuleScope` API.

hubertp committed:
ModuleScope.reset() no more

New builder means new builder (minus types).

hubertp committed:
cleanup

hubertp committed:
drop aux methods by delegating to proxy

hubertp committed:
Ensure elements of ModuleScope are immutable

hubertp committed:
Merge branch 'develop' into wip/hubert/9736-visualizations

4e6 commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java:
```diff
@@ -460,4 +261,265 @@ Type getType() {
+                  Collections.unmodifiableSet(imports),
+                  Collections.unmodifiableSet(exports));
+        }
+        return proxy;
```
From the method description, I thought that `asModuleScope` purpose is to make the builder re-usable by returning a _view_ on its current state.
But in fact, it is not a _view_ of the current builder's state, but a _snapshot_ taken at some point because the `proxy` is stored within the builder. I.e.
```java
var view1 = builder.asModuleScope();
builder.registerType(t);
var view2 = builder.asModuleScope();
view1 == view2; // true, but I'd expect it to be false
```

This behavior can be unexpected and lead to incorrect usage of the method.

hubertp replied:
That is true. I was mostly concerned about having too many of such short-lived _views_. We typically get a module or an associated type and throw away the scope. But maybe it was the case of _premature_ optimization.

hubertp replied:
Done

JaroslavTulach replied:
> Done

Where? In 6b4a208624bdb97a4beaa863f2f105df4467be38 probably...

JaroslavTulach commented on distribution/lib/Standard/Base/0.0.0-dev/src/Meta.enso:
```diff
@@ -348,7 +340,7 @@ get_polyglot_language value = @Builtin_Method "Meta.get_polyglot_language"
-   - scope: The scope in which the symbol name is unresolved.
+   - symbol: The unresolved symbol from which to get the scope.
 create_unresolved_symbol : Text -> Any -> Unresolved_Symbol
 create_unresolved_symbol name scope = @Builtin_Method "Meta.create_unresolved_symbol"
```
The argument is still named `scope`.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java:
```diff
@@ -458,21 +456,21 @@ void unsafeSetIr(org.enso.compiler.core.ir.Module ir) {
-    } else {
-      return scope.withTypes(types);
-    }
+  public ModuleScope.Builder getScopeBuilder() {
```
Exposing _current builder_ (created by whoever else), isn't really safe.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java:
```diff
@@ -458,21 +456,21 @@ void unsafeSetIr(org.enso.compiler.core.ir.Module ir) {
+    return scopeBuilder;
+  }
+
+  public ModuleScope.Builder newScopeBuilder() {
```
Creating _new builder_ is OK. But storing it in `this.scopeBuilder` variable isn't correct.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java:
```diff
@@ -458,21 +456,21 @@ void unsafeSetIr(org.enso.compiler.core.ir.Module ir) {
+    return builder;
+  }
+
+  public void resetScope() {
```
I'd prefer `newScopeBuilder(boolean inheritTypes)` slightly...

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java:
```diff
@@ -53,18 +53,18 @@
 public final class Module implements EnsoObject {
-  private ModuleScope scope;
   private ModuleSources sources;
+  private QualifiedName name;
```
Why isn't `QualifiedName` `final`? I don't see any write to it...

hubertp replied:
Three is a `rename` method.

JaroslavTulach replied:
I see. Thank you. However, in the name of exorcising mutability away from `Module` it may be reasonable to ask whether _name_ shouldn't be a _temporary property_ - e.g. be held by `ModuleScope`?

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java:
```diff
@@ -53,18 +53,18 @@
+  private ModuleScope.Builder scopeBuilder;
+  private final Package<TruffleFile> pkg;
+  private final Cache<ModuleCache.CachedModule, ModuleCache.Metadata> cache;
+  private boolean wasLoadedFromCache;
```
Make `final`?

hubertp replied:
No, there is a setter for it.

JaroslavTulach replied:
Maybe it should be a flag on `ModuleScope`, rather than `Module` itself then! The _immutable content_ of the `Module` is held by its _scope_ and whether or not it was read from a cache is a property of the _content_, not of the module which mutates.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/EnsoRootNode.java:
```diff
@@ -149,6 +147,10 @@ public ModuleScope getModuleScope() {
     return moduleScope;
```
Is it correct that `getModuleScope()` remains here? Shouldn't we have just `getModule().getModuleScope()` - or is it expected that when a `ModuleScope` changes a new `EnsoRootNode` is created?

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/EnsoLanguage.java:
```diff
@@ -265,7 +265,10 @@ protected ExecutableNode parse(InlineParsingRequest request) throws InlineParsin
-                  context, request.getSource(), m.getScope(), redirectConfigWithStrictErrors);
+                  context,
+                  request.getSource(),
+                  m.getScopeBuilder(),
```
This should request `newScopeBuilder()`, fill it and _build it_ (install it) to the `Module`.

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/context/CompilerContext.java:
```diff
@@ -147,5 +148,11 @@ public abstract static class Module {
 
     public abstract boolean isPrivate();
+
+    public abstract ModuleScopeBuilder getScopeBuilder();
```
Removing `getScopeBuilder()` would make the system more predictable.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/atom/AtomConstructor.java:
```diff
@@ -43,7 +43,7 @@
 
   private final String name;
-  private final ModuleScope definitionScope;
+  private final ModuleScope.Builder definitionScope;
```
`AtomConstructor` belongs to `Type` and `Type` shall belong to `Module`.  We should have `Module` here and obtain scope from it.

hubertp replied:
That one is actually doable, despite requiring builder for registering qualified accessor. I will push a change.

hubertp replied:
That one is actually doable, despite requiring builder for registering qualified accessor.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Builtins.java:
```diff
@@ -126,19 +126,20 @@ public static class Debug {
     module = Module.empty(QualifiedName.fromString(MODULE_NAME), null);
-    scope = module.compileScope(context);
+    module.compileScope(context); // Dummy compilation for an empty module
+    ModuleScope.Builder scopeBuilder = module.newScopeBuilder();
```
Using `newScopeBuilder()` seems OK.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ImportExportScope.java:
```diff
@@ -0,0 +1,70 @@
+ * A proxy scope delegating to the underlying module's scope. Additionally, `ImportExportScope` may
+ * limit the number of types that are imported/exported.
+ */
+public class ImportExportScope implements EnsoObject {
```
Why don't this object extend `EnsoObject`? It shouldn't flow thru user data right?

Make `final`.

hubertp replied:
? `EnsoObject` is an interface, I don't understand the request.

JaroslavTulach replied:
Yes, `EnsoObject` is an interface with a _purpose_ (described in its javadoc) of:
```
/** All non-primitive Enso types extends from {@code EnsoObject}. */
```
why do you believe `ImportExportScope` should be a **non-primitive Enso type**? Do you believe this object shall _flow thru user program_? Unless I am mistaken you have removed exposure of `ModuleScope` from `Meta.enso` claiming _"it is dangerous"_ - why do you believe `ImportExportScope` should be exposed (and manipulated) by Enso programs?

```suggestion
public final class ImportExportScope {
```

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/TruffleCompilerModuleScopeBuilder.java:
```diff
@@ -0,0 +1,16 @@
+
+import org.enso.compiler.context.CompilerContext;
+
+public class TruffleCompilerModuleScopeBuilder extends CompilerContext.ModuleScopeBuilder {
```
`final`

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/TruffleCompilerContext.java:
```diff
@@ -723,8 +734,7 @@ public void close() {
       if (resetScope) {
-        module.module.ensureScopeExists();
-        module.module.getScope().reset();
+        module.module.resetScope();
```
I would expect this to call `newBuilder()`.

hubertp replied:
`resetScope` does call `newBuilder`

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/debug/EvalNode.java:
```diff
@@ -86,7 +86,7 @@ RootCallTarget parseExpression(LocalScope scope, ModuleScope moduleScope, String
     var mod = newInlineContext.getModule();
     var m = org.enso.interpreter.runtime.Module.fromCompilerModule(mod);
-    var toTruffle = new IrToTruffle(context, src, m.getScope(), compiler.getConfig());
+    var toTruffle = new IrToTruffle(context, src, m.getScopeBuilder(), compiler.getConfig());
```
I can imagine that this is a place where `EvalNode` wants to incrementally add something to the scope. Still having some sequence like:
```
var b = m.newScopeBuilder(reuseEverything=true)
b.registerXyz(...)
b.build() // and install into `Module`
```
is a way better than having a getter to existing scope builder.

JaroslavTulach reviewed (approved):
I don't think the PR gets us to a state where I'd like to be, but it is certainly making the system better. Ideally I want:
```java
var b = module.newScopeBuilder(reuseYesOrNo);
b.register(...)
b.build() // andInstall();
```
where `buildAndInstall()` actually verifies the current module scope is _the same_ as the scope when the builder was created and otherwise yields an inconsistency error.

hubertp committed:
Adapt tests

hubertp committed:
nit

hubertp committed:
PR review

hubertp committed:
Type can only have a reference to Module

Akirathan commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/dispatch/InvokeFunctionNode.java:
```diff
@@ -276,33 +275,22 @@ private boolean isInSameProject(Function function) {
-      if (modScope != null) {
-        return modScope.getModule().getPackage();
-      }
+      return thisRootNode.getModuleScope().getModule().getPackage();
```
If `thisRootNode.getModuleScope()` is guaranteed to not return `null`, please, also add `Objects.requireNonNull(moduleScope)` to the constructor of `EnsoRootNode`.

hubertp replied:
I thought so as well but apparently this is broken somewhere with autoscoping (even before my changes). I will file a separate issue.

hubertp replied:
https://github.com/enso-org/enso/issues/10186

Akirathan commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java:
```diff
@@ -256,59 +127,49 @@ public Function lookupConversionDefinition(Type original, Type target) {
-  public void addExport(ModuleScope scope) {
-    exports.add(scope);
+  public List<Type> getAllTypes(String name) {
+    var tpes = new LinkedList<Type>();
```
minor: What about `new ArrayList<Type>(types.values().size() + 1)`?

hubertp committed:
Drop caching of views

Akirathan commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/atom/AtomConstructor.java:
```diff
@@ -63,8 +64,8 @@ public final class AtomConstructor implements EnsoObject {
    * @param definitionScope the scope in which this constructor was defined
```
nitpick: since javadoc

Akirathan commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java:
```diff
@@ -460,4 +261,258 @@ Type getType() {
+    @CompilerDirectives.CompilationFinal private ModuleScope moduleScope = null;
+    private final Module module;
+    private final Type associatedType;
+    private Map<String, Object> polyglotSymbols;
```
nitpick: fields `polyglotSymbols`, `type`, `methods`, `conversions`, `imports` and `exports` can all be final

Akirathan commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/job/UpsertVisualizationJob.scala:
```diff
@@ -63,70 +63,66 @@ class UpsertVisualizationJob(
-                  Api.Response(Api.ModuleNotFound(moduleName))
-                )
-                None
+      () => {
```
How are the changes in this source related to module scope builder?

hubertp replied:
We can now reduce the scope of the lock, which was previously experiencing a race condition because of mutable `ModuleScope`.

Akirathan reviewed (approved):
Generally looks like a better design than the one we currently have. There might be some follow-up issues.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java:
```diff
@@ -460,4 +261,258 @@ Type getType() {
+
+  public static class Builder {
+
+    @CompilerDirectives.CompilationFinal private ModuleScope moduleScope = null;
```
I doubt this is correct. How is a code using this `moduleScope` field in PE going to be invalidated when the field changes?

Anyway, I believe `Builder` shouldn't be used on _fast path_. Rather put `CompilerAsserts.neverPartOfCompilation();` to every method of the `Builder`!

hubertp committed:
PR comments

hubertp committed:
Merge branch 'develop' into wip/hubert/9736-visualizations

hubertp committed:
fix compilation issue

hubertp commented:
There are few minor issues that I'd like to address separately, as this PR has already been dragging on. Most importantly remove the `getScopeBuilder` method and instead pass it around during compilation. But I will make the change in a follow up.

hubertp committed:
nit
